--- a/drivers/usb/gadget/webcam.c
+++ b/drivers/usb/gadget/webcam.c
@@ -17,7 +17,6 @@
 #include <linux/usb/video.h>
 
 #include "f_uvc.h"
-#include "f_uac_mic.h"
 
 /*
  * Kbuild is not very cooperative with respect to linking separately
@@ -30,7 +29,6 @@
 #include "uvc_video.c"
 #include "uvc_v4l2.c"
 #include "f_uvc.c"
-#include "f_uac_mic.c"
 
 USB_GADGET_COMPOSITE_OPTIONS();
 
@@ -45,21 +43,24 @@
 static char webcam_vendor_label[] = "thingino";
 static char webcam_product_label[] = "Raptor Webcam";
 static char webcam_config_label[] = "Video";
+static char webcam_extension_label[] = "Thingino Camera Controls";
 
 /* string IDs are assigned dynamically */
 
 #define STRING_DESCRIPTION_IDX		USB_GADGET_FIRST_AVAIL_IDX
+#define STRING_EXTENSION_IDX		(STRING_DESCRIPTION_IDX + 1)
 
 static struct usb_string webcam_strings[] = {
 	[USB_GADGET_MANUFACTURER_IDX].s = webcam_vendor_label,
 	[USB_GADGET_PRODUCT_IDX].s = webcam_product_label,
 	[USB_GADGET_SERIAL_IDX].s = "",
 	[STRING_DESCRIPTION_IDX].s = webcam_config_label,
+	[STRING_EXTENSION_IDX].s = webcam_extension_label,
 	{  }
 };
 
 static struct usb_gadget_strings webcam_stringtab = {
-	.language = 0x0409,	/* en-us */
+	.language = 0x0409,
 	.strings = webcam_strings,
 };
 
@@ -127,11 +128,27 @@
 	.bSourceID		= 1,
 	.wMaxMultiplier		= cpu_to_le16(16*1024),
 	.bControlSize		= 2,
-	.bmControls[0]		= 0,
-	.bmControls[1]		= 0,
+	.bmControls[0]		= (1 << 0) | (1 << 1) | (1 << 2) | (1 << 3) | (1 << 4) | (1 << 5) | (1 << 6),
+	.bmControls[1]		= (1 << 0) | (1 << 2) | (1 << 4),
 	.iProcessing		= 0,
 };
 
+DECLARE_UVC_EXTENSION_UNIT_DESCRIPTOR(1, 2);
+static struct UVC_EXTENSION_UNIT_DESCRIPTOR(1, 2) uvc_extension = {
+	.bLength = UVC_DT_EXTENSION_UNIT_SIZE(1, 2),
+	.bDescriptorType = USB_DT_CS_INTERFACE,
+	.bDescriptorSubType = UVC_VC_EXTENSION_UNIT,
+	.bUnitID = 4,
+	.guidExtensionCode = { 0x54, 0x48, 0x49, 0x4e, 0x47, 0x49, 0x4e, 0x4f,
+		0x2d, 0x55, 0x56, 0x43, 0x44, 0x2d, 0x01, 0x00 },
+	.bNumControls = 12,
+	.bNrInPins = 1,
+	.baSourceID = { 2 },
+	.bControlSize = 2,
+	.bmControls = { 0xff, 0x0f },
+	.iExtension = 0,
+};
+
 static const struct uvc_output_terminal_descriptor uvc_output_terminal = {
 	.bLength		= UVC_DT_OUTPUT_TERMINAL_SIZE,
 	.bDescriptorType	= USB_DT_CS_INTERFACE,
@@ -173,7 +190,7 @@
 	.bDescriptorType	= USB_DT_CS_INTERFACE,
 	.bDescriptorSubType	= UVC_VS_FORMAT_MJPEG,
 	.bFormatIndex		= 1,
-	.bNumFrameDescriptors	= 2,
+	.bNumFrameDescriptors	= 6,
 	.bmFlags		= 0,
 	.bDefaultFrameIndex	= 1,
 	.bAspectRatioX		= 0,
@@ -182,10 +199,10 @@
 	.bCopyProtect		= 0,
 };
 
-DECLARE_UVC_FRAME_MJPEG(3);
+DECLARE_UVC_FRAME_MJPEG(6);
 
-static const struct UVC_FRAME_MJPEG(3) uvc_frame_mjpg_1080p = {
-	.bLength		= UVC_DT_FRAME_MJPEG_SIZE(3),
+static const struct UVC_FRAME_MJPEG(6) uvc_frame_mjpg_1080p = {
+	.bLength		= UVC_DT_FRAME_MJPEG_SIZE(6),
 	.bDescriptorType	= USB_DT_CS_INTERFACE,
 	.bDescriptorSubType	= UVC_VS_FRAME_MJPEG,
 	.bFrameIndex		= 1,
@@ -202,11 +219,11 @@
 	.dwFrameInterval[2]	= cpu_to_le32(666666),	/* 15 fps */
 };
 
-static const struct UVC_FRAME_MJPEG(3) uvc_frame_mjpg_360p = {
-	.bLength		= UVC_DT_FRAME_MJPEG_SIZE(3),
+static const struct UVC_FRAME_MJPEG(6) uvc_frame_mjpg_360p = {
+	.bLength		= UVC_DT_FRAME_MJPEG_SIZE(6),
 	.bDescriptorType	= USB_DT_CS_INTERFACE,
 	.bDescriptorSubType	= UVC_VS_FRAME_MJPEG,
-	.bFrameIndex		= 2,
+	.bFrameIndex		= 6,
 	.bmCapabilities		= 0,
 	.wWidth			= cpu_to_le16(640),
 	.wHeight		= cpu_to_le16(360),
@@ -220,6 +237,46 @@
 	.dwFrameInterval[2]	= cpu_to_le32(666666),	/* 15 fps */
 };
 
+static const struct UVC_FRAME_MJPEG(6) uvc_frame_mjpg_1280x960 = {
+	.bLength = UVC_DT_FRAME_MJPEG_SIZE(6), .bDescriptorType = USB_DT_CS_INTERFACE,
+	.bDescriptorSubType = UVC_VS_FRAME_MJPEG, .bFrameIndex = 2, .bmCapabilities = 0,
+	.wWidth = cpu_to_le16(1280), .wHeight = cpu_to_le16(960),
+	.dwMinBitRate = cpu_to_le32(2000000), .dwMaxBitRate = cpu_to_le32(30000000),
+	.dwMaxVideoFrameBufferSize = cpu_to_le32(1280 * 960 * 2),
+	.dwDefaultFrameInterval = cpu_to_le32(333333), .bFrameIntervalType = 3,
+	.dwFrameInterval = { cpu_to_le32(333333), cpu_to_le32(400000), cpu_to_le32(666666) },
+};
+
+static const struct UVC_FRAME_MJPEG(6) uvc_frame_mjpg_720p = {
+	.bLength = UVC_DT_FRAME_MJPEG_SIZE(6), .bDescriptorType = USB_DT_CS_INTERFACE,
+	.bDescriptorSubType = UVC_VS_FRAME_MJPEG, .bFrameIndex = 3, .bmCapabilities = 0,
+	.wWidth = cpu_to_le16(1280), .wHeight = cpu_to_le16(720),
+	.dwMinBitRate = cpu_to_le32(1500000), .dwMaxBitRate = cpu_to_le32(25000000),
+	.dwMaxVideoFrameBufferSize = cpu_to_le32(1280 * 720 * 2),
+	.dwDefaultFrameInterval = cpu_to_le32(333333), .bFrameIntervalType = 3,
+	.dwFrameInterval = { cpu_to_le32(333333), cpu_to_le32(400000), cpu_to_le32(666666) },
+};
+
+static const struct UVC_FRAME_MJPEG(6) uvc_frame_mjpg_800x600 = {
+	.bLength = UVC_DT_FRAME_MJPEG_SIZE(6), .bDescriptorType = USB_DT_CS_INTERFACE,
+	.bDescriptorSubType = UVC_VS_FRAME_MJPEG, .bFrameIndex = 4, .bmCapabilities = 0,
+	.wWidth = cpu_to_le16(800), .wHeight = cpu_to_le16(600),
+	.dwMinBitRate = cpu_to_le32(1000000), .dwMaxBitRate = cpu_to_le32(18000000),
+	.dwMaxVideoFrameBufferSize = cpu_to_le32(800 * 600 * 2),
+	.dwDefaultFrameInterval = cpu_to_le32(333333), .bFrameIntervalType = 3,
+	.dwFrameInterval = { cpu_to_le32(333333), cpu_to_le32(400000), cpu_to_le32(666666) },
+};
+
+static const struct UVC_FRAME_MJPEG(6) uvc_frame_mjpg_640x480 = {
+	.bLength = UVC_DT_FRAME_MJPEG_SIZE(6), .bDescriptorType = USB_DT_CS_INTERFACE,
+	.bDescriptorSubType = UVC_VS_FRAME_MJPEG, .bFrameIndex = 5, .bmCapabilities = 0,
+	.wWidth = cpu_to_le16(640), .wHeight = cpu_to_le16(480),
+	.dwMinBitRate = cpu_to_le32(800000), .dwMaxBitRate = cpu_to_le32(12000000),
+	.dwMaxVideoFrameBufferSize = cpu_to_le32(640 * 480 * 2),
+	.dwDefaultFrameInterval = cpu_to_le32(333333), .bFrameIntervalType = 3,
+	.dwFrameInterval = { cpu_to_le32(333333), cpu_to_le32(400000), cpu_to_le32(666666) },
+};
+
 /* --- Format 2: H.264 (Frame-Based) --- */
 
 /*
@@ -264,14 +321,14 @@
 		__u32 dwFrameInterval[];			\
 	} __attribute__((packed))
 
-DECLARE_UVC_FRAME_FRAMEBASED(3);
+DECLARE_UVC_FRAME_FRAMEBASED(6);
 
 static const struct uvc_format_framebased uvc_format_h264 = {
 	.bLength		= UVC_DT_FORMAT_FRAMEBASED_SIZE,
 	.bDescriptorType	= USB_DT_CS_INTERFACE,
 	.bDescriptorSubType	= UVC_VS_FORMAT_FRAME_BASED,
 	.bFormatIndex		= 2,
-	.bNumFrameDescriptors	= 2,
+	.bNumFrameDescriptors	= 6,
 	.guidFormat		=
 		{ 'H',  '2',  '6',  '4', 0x00, 0x00, 0x10, 0x00,
 		 0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71},
@@ -284,8 +341,8 @@
 	.bVariableSize		= 1,
 };
 
-static const struct UVC_FRAME_FRAMEBASED_3 uvc_frame_h264_1080p = {
-	.bLength		= UVC_DT_FRAME_FRAMEBASED_SIZE(3),
+static const struct UVC_FRAME_FRAMEBASED_6 uvc_frame_h264_1080p = {
+	.bLength		= UVC_DT_FRAME_FRAMEBASED_SIZE(6),
 	.bDescriptorType	= USB_DT_CS_INTERFACE,
 	.bDescriptorSubType	= UVC_VS_FRAME_FRAME_BASED,
 	.bFrameIndex		= 1,
@@ -304,11 +361,11 @@
 	},
 };
 
-static const struct UVC_FRAME_FRAMEBASED_3 uvc_frame_h264_360p = {
-	.bLength		= UVC_DT_FRAME_FRAMEBASED_SIZE(3),
+static const struct UVC_FRAME_FRAMEBASED_6 uvc_frame_h264_360p = {
+	.bLength		= UVC_DT_FRAME_FRAMEBASED_SIZE(6),
 	.bDescriptorType	= USB_DT_CS_INTERFACE,
 	.bDescriptorSubType	= UVC_VS_FRAME_FRAME_BASED,
-	.bFrameIndex		= 2,
+	.bFrameIndex		= 6,
 	.bmCapabilities		= 0,
 	.wWidth			= cpu_to_le16(640),
 	.wHeight		= cpu_to_le16(360),
@@ -324,6 +381,42 @@
 	},
 };
 
+static const struct UVC_FRAME_FRAMEBASED_6 uvc_frame_h264_1280x960 = {
+	.bLength = UVC_DT_FRAME_FRAMEBASED_SIZE(6), .bDescriptorType = USB_DT_CS_INTERFACE,
+	.bDescriptorSubType = UVC_VS_FRAME_FRAME_BASED, .bFrameIndex = 2, .bmCapabilities = 0,
+	.wWidth = cpu_to_le16(1280), .wHeight = cpu_to_le16(960),
+	.dwMinBitRate = cpu_to_le32(400000), .dwMaxBitRate = cpu_to_le32(6000000),
+	.dwDefaultFrameInterval = cpu_to_le32(333333), .bFrameIntervalType = 3, .dwBytesPerLine = 0,
+	.dwFrameInterval = { cpu_to_le32(333333), cpu_to_le32(400000), cpu_to_le32(666666) },
+};
+
+static const struct UVC_FRAME_FRAMEBASED_6 uvc_frame_h264_720p = {
+	.bLength = UVC_DT_FRAME_FRAMEBASED_SIZE(6), .bDescriptorType = USB_DT_CS_INTERFACE,
+	.bDescriptorSubType = UVC_VS_FRAME_FRAME_BASED, .bFrameIndex = 3, .bmCapabilities = 0,
+	.wWidth = cpu_to_le16(1280), .wHeight = cpu_to_le16(720),
+	.dwMinBitRate = cpu_to_le32(300000), .dwMaxBitRate = cpu_to_le32(5000000),
+	.dwDefaultFrameInterval = cpu_to_le32(333333), .bFrameIntervalType = 3, .dwBytesPerLine = 0,
+	.dwFrameInterval = { cpu_to_le32(333333), cpu_to_le32(400000), cpu_to_le32(666666) },
+};
+
+static const struct UVC_FRAME_FRAMEBASED_6 uvc_frame_h264_800x600 = {
+	.bLength = UVC_DT_FRAME_FRAMEBASED_SIZE(6), .bDescriptorType = USB_DT_CS_INTERFACE,
+	.bDescriptorSubType = UVC_VS_FRAME_FRAME_BASED, .bFrameIndex = 4, .bmCapabilities = 0,
+	.wWidth = cpu_to_le16(800), .wHeight = cpu_to_le16(600),
+	.dwMinBitRate = cpu_to_le32(200000), .dwMaxBitRate = cpu_to_le32(3500000),
+	.dwDefaultFrameInterval = cpu_to_le32(333333), .bFrameIntervalType = 3, .dwBytesPerLine = 0,
+	.dwFrameInterval = { cpu_to_le32(333333), cpu_to_le32(400000), cpu_to_le32(666666) },
+};
+
+static const struct UVC_FRAME_FRAMEBASED_6 uvc_frame_h264_640x480 = {
+	.bLength = UVC_DT_FRAME_FRAMEBASED_SIZE(6), .bDescriptorType = USB_DT_CS_INTERFACE,
+	.bDescriptorSubType = UVC_VS_FRAME_FRAME_BASED, .bFrameIndex = 5, .bmCapabilities = 0,
+	.wWidth = cpu_to_le16(640), .wHeight = cpu_to_le16(480),
+	.dwMinBitRate = cpu_to_le32(150000), .dwMaxBitRate = cpu_to_le32(3000000),
+	.dwDefaultFrameInterval = cpu_to_le32(333333), .bFrameIntervalType = 3, .dwBytesPerLine = 0,
+	.dwFrameInterval = { cpu_to_le32(333333), cpu_to_le32(400000), cpu_to_le32(666666) },
+};
+
 /* --- Color matching --- */
 
 static const struct uvc_color_matching_descriptor uvc_color_matching = {
@@ -343,6 +436,7 @@
 	(const struct uvc_descriptor_header *) &uvc_control_header,
 	(const struct uvc_descriptor_header *) &uvc_camera_terminal,
 	(const struct uvc_descriptor_header *) &uvc_processing,
+	(const struct uvc_descriptor_header *) &uvc_extension,
 	(const struct uvc_descriptor_header *) &uvc_output_terminal,
 	NULL,
 };
@@ -351,6 +445,7 @@
 	(const struct uvc_descriptor_header *) &uvc_control_header,
 	(const struct uvc_descriptor_header *) &uvc_camera_terminal,
 	(const struct uvc_descriptor_header *) &uvc_processing,
+	(const struct uvc_descriptor_header *) &uvc_extension,
 	(const struct uvc_descriptor_header *) &uvc_output_terminal,
 	NULL,
 };
@@ -359,9 +454,17 @@
 	(const struct uvc_descriptor_header *) &uvc_input_header,
 	(const struct uvc_descriptor_header *) &uvc_format_mjpg,
 	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_1080p,
+	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_1280x960,
+	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_720p,
+	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_800x600,
+	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_640x480,
 	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_360p,
 	(const struct uvc_descriptor_header *) &uvc_format_h264,
 	(const struct uvc_descriptor_header *) &uvc_frame_h264_1080p,
+	(const struct uvc_descriptor_header *) &uvc_frame_h264_1280x960,
+	(const struct uvc_descriptor_header *) &uvc_frame_h264_720p,
+	(const struct uvc_descriptor_header *) &uvc_frame_h264_800x600,
+	(const struct uvc_descriptor_header *) &uvc_frame_h264_640x480,
 	(const struct uvc_descriptor_header *) &uvc_frame_h264_360p,
 	(const struct uvc_descriptor_header *) &uvc_color_matching,
 	NULL,
@@ -371,9 +474,17 @@
 	(const struct uvc_descriptor_header *) &uvc_input_header,
 	(const struct uvc_descriptor_header *) &uvc_format_mjpg,
 	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_1080p,
+	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_1280x960,
+	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_720p,
+	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_800x600,
+	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_640x480,
 	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_360p,
 	(const struct uvc_descriptor_header *) &uvc_format_h264,
 	(const struct uvc_descriptor_header *) &uvc_frame_h264_1080p,
+	(const struct uvc_descriptor_header *) &uvc_frame_h264_1280x960,
+	(const struct uvc_descriptor_header *) &uvc_frame_h264_720p,
+	(const struct uvc_descriptor_header *) &uvc_frame_h264_800x600,
+	(const struct uvc_descriptor_header *) &uvc_frame_h264_640x480,
 	(const struct uvc_descriptor_header *) &uvc_frame_h264_360p,
 	(const struct uvc_descriptor_header *) &uvc_color_matching,
 	NULL,
@@ -383,9 +494,17 @@
 	(const struct uvc_descriptor_header *) &uvc_input_header,
 	(const struct uvc_descriptor_header *) &uvc_format_mjpg,
 	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_1080p,
+	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_1280x960,
+	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_720p,
+	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_800x600,
+	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_640x480,
 	(const struct uvc_descriptor_header *) &uvc_frame_mjpg_360p,
 	(const struct uvc_descriptor_header *) &uvc_format_h264,
 	(const struct uvc_descriptor_header *) &uvc_frame_h264_1080p,
+	(const struct uvc_descriptor_header *) &uvc_frame_h264_1280x960,
+	(const struct uvc_descriptor_header *) &uvc_frame_h264_720p,
+	(const struct uvc_descriptor_header *) &uvc_frame_h264_800x600,
+	(const struct uvc_descriptor_header *) &uvc_frame_h264_640x480,
 	(const struct uvc_descriptor_header *) &uvc_frame_h264_360p,
 	(const struct uvc_descriptor_header *) &uvc_color_matching,
 	NULL,
@@ -397,8 +516,6 @@
 
 static struct usb_function_instance *fi_acm;
 static struct usb_function *f_acm;
-static struct usb_function_instance *fi_ncm;
-static struct usb_function *f_ncm;
 
 static int __init
 webcam_config_bind(struct usb_configuration *c)
@@ -411,10 +528,6 @@
 	if (ret < 0)
 		return ret;
 
-	ret = uac_mic_bind_config(c);
-	if (ret < 0)
-		pr_warn("g_webcam: UAC mic bind failed: %d\n", ret);
-
 	if (fi_acm) {
 		f_acm = usb_get_function(fi_acm);
 		if (IS_ERR(f_acm)) {
@@ -428,19 +541,6 @@
 		}
 	}
 
-	if (fi_ncm) {
-		f_ncm = usb_get_function(fi_ncm);
-		if (IS_ERR(f_ncm)) {
-			pr_warn("g_webcam: NCM get_function failed: %ld\n",
-				PTR_ERR(f_ncm));
-			f_ncm = NULL;
-		} else {
-			ret = usb_add_function(c, f_ncm);
-			if (ret < 0)
-				pr_warn("g_webcam: NCM add_function failed: %d\n", ret);
-		}
-	}
-
 	return 0;
 }
 
@@ -455,10 +555,6 @@
 static int /* __init_or_exit */
 webcam_unbind(struct usb_composite_dev *cdev)
 {
-	if (f_ncm)
-		usb_put_function(f_ncm);
-	if (fi_ncm)
-		usb_put_function_instance(fi_ncm);
 	if (f_acm)
 		usb_put_function(f_acm);
 	if (fi_acm)
@@ -480,6 +576,7 @@
 		webcam_strings[USB_GADGET_PRODUCT_IDX].id;
 	webcam_config_driver.iConfiguration =
 		webcam_strings[STRING_DESCRIPTION_IDX].id;
+	uvc_extension.iExtension = webcam_strings[STRING_EXTENSION_IDX].id;
 
 	/* ACM serial: get function instance from f_acm module */
 	fi_acm = usb_get_function_instance("acm");
@@ -489,14 +586,6 @@
 		fi_acm = NULL;
 	}
 
-	/* NCM ethernet: get function instance from f_ncm module */
-	fi_ncm = usb_get_function_instance("ncm");
-	if (IS_ERR(fi_ncm)) {
-		pr_warn("g_webcam: NCM function not available: %ld\n",
-			PTR_ERR(fi_ncm));
-		fi_ncm = NULL;
-	}
-
 	if ((ret = usb_add_config(cdev, &webcam_config_driver,
 					webcam_config_bind)) < 0)
 		goto error;
